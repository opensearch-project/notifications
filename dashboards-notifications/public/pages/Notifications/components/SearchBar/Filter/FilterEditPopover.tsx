/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  EuiButton,
  EuiButtonEmpty,
  EuiComboBox,
  EuiComboBoxOptionOption,
  EuiFlexGroup,
  EuiFlexItem,
  EuiFormRow,
  EuiSpacer,
} from '@elastic/eui';
import React, { useContext, useMemo, useState } from 'react';
import { MainContext } from '../../../../Main/Main';
import {
  FilterOperatorType,
  FilterFieldType,
  getFilterOperatorOptions,
  getOperatorString,
  getFilterFieldOptions,
  getValueComponent,
} from '../utils/filterHelpers';
import { FilterType } from './Filters';

interface FilterEditPopoverProps {
  filter?: FilterType;
  closePopover: () => void;
  index?: number;
  setFilter: (newFilter: FilterType, index?: number) => void;
}

export function FilterEditPopover(props: FilterEditPopoverProps) {
  const mainStateContext = useContext(MainContext)!;
  const [selectedFieldOptions, setSelectedFieldOptions] = useState<
    Array<EuiComboBoxOptionOption<string>>
  >(props.filter ? [{ label: props.filter.field }] : []);
  const [selectedOperatorOptions, setSelectedOperatorOptions] = useState<
    Array<EuiComboBoxOptionOption<string>>
  >(
    props.filter
      ? [
          {
            label: getOperatorString(
              props.filter.operator,
              props.filter.inverted
            ),
          },
        ]
      : []
  );
  const [filterValue, setFilterValue] = useState<FilterType['value']>(
    props.filter?.value || []
  );

  const filterFieldOptions = useMemo(getFilterFieldOptions, []);

  return (
    <div style={{ width: 400 }}>
      {/* invisible button workaround to prevent auto focus on context menu panel switch */}
      <button
        style={{
          width: 0,
          height: 0,
          position: 'fixed',
          marginLeft: -10000,
          bottom: 0,
        }}
      />
      <EuiFlexGroup gutterSize="s">
        <EuiFlexItem grow={6}>
          <EuiFormRow label="Field">
            <EuiComboBox
              placeholder="Select a field first"
              isClearable={false}
              options={filterFieldOptions}
              selectedOptions={selectedFieldOptions}
              onChange={(e) => {
                setSelectedFieldOptions(e);
                setSelectedOperatorOptions([]);
                setFilterValue(null);
              }}
              singleSelection={{ asPlainText: true }}
            />
          </EuiFormRow>
        </EuiFlexItem>
        <EuiFlexItem grow={5}>
          <EuiFormRow label="Operator">
            <EuiComboBox
              placeholder={
                selectedFieldOptions.length === 0 ? 'Waiting' : 'Select'
              }
              isClearable={false}
              isDisabled={selectedFieldOptions.length === 0}
              options={
                selectedFieldOptions.length === 0
                  ? []
                  : getFilterOperatorOptions(
                      selectedFieldOptions[0].label as FilterFieldType
                    )
              }
              selectedOptions={selectedOperatorOptions}
              onChange={(e) => {
                setSelectedOperatorOptions(e);
                setFilterValue(null);
              }}
              singleSelection={{ asPlainText: true }}
            />
          </EuiFormRow>
        </EuiFlexItem>
      </EuiFlexGroup>
      {selectedOperatorOptions.length > 0 &&
        getValueComponent(
          selectedFieldOptions[0].label as FilterFieldType,
          selectedOperatorOptions[0].label as FilterOperatorType,
          filterValue,
          setFilterValue,
          mainStateContext.availableChannels
        )}
      <EuiSpacer size="m" />
      <EuiFlexGroup gutterSize="s" justifyContent="flexEnd">
        <EuiFlexItem grow={false}>
          <EuiButtonEmpty
            data-test-subj="filter-popover-cancel-button"
            onClick={props.closePopover}
          >
            Cancel
          </EuiButtonEmpty>
        </EuiFlexItem>
        <EuiFlexItem grow={false}>
          <EuiButton
            fill
            disabled={
              selectedFieldOptions.length === 0 ||
              selectedOperatorOptions.length === 0 ||
              !(filterValue && filterValue.length > 0)
            }
            onClick={() => {
              props.closePopover();
              props.setFilter(
                {
                  field: selectedFieldOptions[0].label as FilterFieldType,
                  operator: selectedOperatorOptions[0]
                    .label as FilterOperatorType,
                  value: filterValue,
                  inverted: selectedOperatorOptions[0].label.includes('not'),
                  disabled: props.filter ? props.filter.disabled : false,
                },
                props.index
              );
            }}
          >
            Save
          </EuiButton>
        </EuiFlexItem>
      </EuiFlexGroup>
    </div>
  );
}
