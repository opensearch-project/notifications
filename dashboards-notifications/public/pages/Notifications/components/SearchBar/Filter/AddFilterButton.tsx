/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { EuiButtonEmpty, EuiPopover, EuiPopoverTitle } from '@elastic/eui';
import React, { useState } from 'react';
import { FilterEditPopover } from './FilterEditPopover';
import { FilterType } from './Filters';

interface AddFilterButtonProps {
  setFilter: (newFilter: FilterType, index?: number) => void;
}

export function AddFilterButton(props: AddFilterButtonProps) {
  const [isPopoverOpen, setIsPopoverOpen] = useState(false);
  const button = (
    <EuiButtonEmpty
      size="xs"
      onClick={() => {
        setIsPopoverOpen(true);
      }}
    >
      + Add filter
    </EuiButtonEmpty>
  );

  return (
    <>
      <EuiPopover
        button={button}
        isOpen={isPopoverOpen}
        closePopover={() => setIsPopoverOpen(false)}
        anchorPosition="downLeft"
        withTitle
      >
        <EuiPopoverTitle>{'Add filter'}</EuiPopoverTitle>
        <FilterEditPopover
          setFilter={props.setFilter}
          closePopover={() => setIsPopoverOpen(false)}
        />
      </EuiPopover>
    </>
  );
}
