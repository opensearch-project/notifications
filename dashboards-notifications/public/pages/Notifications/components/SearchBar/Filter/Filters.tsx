/*
 * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import {
  EuiBadge,
  EuiContextMenu,
  EuiContextMenuPanelDescriptor,
  EuiFlexGroup,
  EuiFlexItem,
  EuiIcon,
  EuiPopover,
  EuiTextColor,
} from '@elastic/eui';
import { EuiComboBoxOptionOption } from '@elastic/eui/src/components/combo_box/types';
import React, { useMemo, useState } from 'react';
import {
  FilterFieldType,
  FilterOperatorType,
  filterToQueryString,
  isSingleSelection,
} from '../utils/filterHelpers';
import { AddFilterButton } from './AddFilterButton';
import { FilterEditPopover } from './FilterEditPopover';
import { GlobalFilterButton } from './GlobalFilterButton';

export interface FilterType {
  field: FilterFieldType;
  operator: FilterOperatorType;
  value: string | Array<EuiComboBoxOptionOption<string>> | null;
  inverted: boolean;
  disabled: boolean;
}

interface FiltersProps {
  filters: FilterType[];
  setFilters: (filters: FilterType[]) => void;
}

export function Filters(props: FiltersProps) {
  // set a filter at an index. if newFilter doesn't exist, remove filter at the index
  // if index doesn't exist, append newFilter to the end
  const setFilter = (newFilter: FilterType | null, index?: number) => {
    if (index === undefined) index = props.filters.length;
    const newFilters = [...props.filters];
    if (newFilter) newFilters.splice(index, 1, newFilter);
    else newFilters.splice(index, 1);
    props.setFilters(newFilters);
  };

  const getFilterPopoverPanels = (
    filter: FilterType,
    index: number,
    closePopover: () => void
  ): EuiContextMenuPanelDescriptor[] => [
    {
      id: 0,
      items: [
        {
          name: 'Edit filter',
          icon: <EuiIcon type="invert" size="m" />,
          panel: 1,
        },
        {
          name: `${filter.inverted ? 'Include' : 'Exclude'} results`,
          icon: (
            <EuiIcon
              type={filter.inverted ? 'plusInCircle' : 'minusInCircle'}
              size="m"
            />
          ),
          onClick: () => {
            filter.inverted = !filter.inverted;
            setFilter(filter, index);
          },
        },
        {
          name: filter.disabled ? 'Re-enable' : 'Temporarily disable',
          icon: (
            <EuiIcon type={filter.disabled ? 'eye' : 'eyeClosed'} size="m" />
          ),
          onClick: () => {
            filter.disabled = !filter.disabled;
            setFilter(filter, index);
          },
        },
        {
          name: 'Delete',
          icon: <EuiIcon type="trash" size="m" />,
          onClick: () => setFilter(null, index),
        },
      ],
    },
    {
      id: 1,
      width: 430,
      title: 'Edit filter',
      content: (
        <div style={{ margin: 15 }}>
          <FilterEditPopover
            filter={filter}
            index={index}
            setFilter={setFilter}
            closePopover={closePopover}
          />
        </div>
      ),
    },
  ];

  const renderFilters = () => {
    const FilterBadge = ({
      filter,
      index,
    }: {
      filter: FilterType;
      index: number;
    }) => {
      const [isPopoverOpen, setIsPopoverOpen] = useState(false);
      const className =
        'globalFilterItem' +
        (filter.disabled ? ' globalFilterItem-isDisabled' : '') +
        (filter.inverted ? ' globalFilterItem-isExcluded' : '');

      let singleSelection = isSingleSelection(filter);
      let value;
      if (typeof filter.value === 'string') {
        value = filter.value;
      } else if (Array.isArray(filter.value)) {
        // combo box array
        value = filter.value.map((option) => option.label).join(', ');
        if (!singleSelection) value = 'is one of ' + value;
      }

      const filterLabel = filter.inverted ? (
        <>
          <EuiTextColor color={filter.disabled ? 'default' : 'danger'}>
            {'NOT '}
          </EuiTextColor>
          <EuiTextColor color="default">{`${filter.field}: ${value}`}</EuiTextColor>
        </>
      ) : (
        `${filter.field}: ${value}`
      );

      const badge = (
        <EuiBadge
          className={className}
          onClick={() => setIsPopoverOpen(true)}
          onClickAriaLabel="Open filter settings"
          color={filter.disabled ? '#e7e9f0' : 'hollow'}
          iconType="cross"
          iconSide="right"
          iconOnClick={() => setFilter(null, index)}
          iconOnClickAriaLabel="Remove filter"
        >
          {filterLabel}
        </EuiBadge>
      );
      return (
        <EuiFlexItem grow={false} key={`filter-${index}`}>
          <EuiPopover
            isOpen={isPopoverOpen}
            closePopover={() => setIsPopoverOpen(false)}
            panelPaddingSize="none"
            button={badge}
          >
            <EuiContextMenu
              initialPanelId={0}
              panels={getFilterPopoverPanels(filter, index, () =>
                setIsPopoverOpen(false)
              )}
            />
          </EuiPopover>
        </EuiFlexItem>
      );
    };

    return (
      <>
        {props.filters.length > 0
          ? props.filters.map((filter, i) => (
              <FilterBadge filter={filter} index={i} key={i} />
            ))
          : null}
      </>
    );
  };

  const filterComponents = useMemo(() => renderFilters(), [props.filters]);

  return (
    <>
      <EuiFlexGroup
        gutterSize="xs"
        alignItems="center"
        responsive={false}
        style={{ minHeight: 32 }}
      >
        <EuiFlexItem grow={false}>
          <GlobalFilterButton
            filters={props.filters}
            setFilters={props.setFilters}
          />
        </EuiFlexItem>
        {filterComponents}
        <EuiFlexItem grow={false}>
          <AddFilterButton setFilter={setFilter} />
        </EuiFlexItem>
      </EuiFlexGroup>
    </>
  );
}
