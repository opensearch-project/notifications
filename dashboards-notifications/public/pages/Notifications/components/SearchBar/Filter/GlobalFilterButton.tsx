/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  EuiButtonIcon,
  EuiContextMenu,
  EuiIcon,
  EuiPopover,
} from '@elastic/eui';
import React, { useState } from 'react';
import { FilterType } from './Filters';

interface GlobalFilterButtonProps {
  filters: FilterType[];
  setFilters: (filters: FilterType[]) => void;
}

export function GlobalFilterButton(props: GlobalFilterButtonProps) {
  const [isPopoverOpen, setIsPopoverOpen] = useState(false);
  const globalPopoverPanels = [
    {
      id: 0,
      title: 'Change all filters',
      items: [
        {
          name: 'Enable all',
          icon: <EuiIcon type="eye" size="m" />,
          onClick: () => {
            props.setFilters(
              props.filters.map((filter) => ({ ...filter, disabled: false }))
            );
            setIsPopoverOpen(false);
          },
        },
        {
          name: 'Disable all',
          icon: <EuiIcon type="eyeClosed" size="m" />,
          onClick: () => {
            props.setFilters(
              props.filters.map((filter) => ({ ...filter, disabled: true }))
            );
            setIsPopoverOpen(false);
          },
        },
        // {
        //   name: 'Invert inclusion',
        //   icon: <EuiIcon type="invert" size="m" />,
        //   onClick: () => {
        //     props.setFilters(
        //       props.filters.map((filter) => ({
        //         ...filter,
        //         inverted: !filter.inverted,
        //       }))
        //     );
        //     setIsPopoverOpen(false);
        //   },
        // },
        {
          name: 'Invert enabled/disabled',
          icon: <EuiIcon type="eye" size="m" />,
          onClick: () => {
            props.setFilters(
              props.filters.map((filter) => ({
                ...filter,
                disabled: !filter.disabled,
              }))
            );
            setIsPopoverOpen(false);
          },
        },
        {
          name: 'Remove all',
          icon: <EuiIcon type="trash" size="m" />,
          onClick: () => {
            props.setFilters([]);
            setIsPopoverOpen(false);
          },
        },
      ],
    },
  ];

  return (
    <EuiPopover
      isOpen={isPopoverOpen}
      closePopover={() => setIsPopoverOpen(false)}
      button={
        <EuiButtonIcon
          onClick={() => setIsPopoverOpen(true)}
          iconType="filter"
          title="Change all filters"
          aria-label="Change all filters"
        />
      }
      anchorPosition="rightUp"
      panelPaddingSize="none"
      withTitle
    >
      <EuiContextMenu initialPanelId={0} panels={globalPopoverPanels} />
    </EuiPopover>
  );
}
