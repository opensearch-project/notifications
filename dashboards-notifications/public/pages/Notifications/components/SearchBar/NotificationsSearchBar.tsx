/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  EuiButton,
  EuiFieldSearch,
  EuiFlexGroup,
  EuiFlexItem,
  EuiSpacer,
  EuiSuperDatePicker,
  ShortDate,
} from '@elastic/eui';
import React, { useState } from 'react';
import { Filters, FilterType } from './Filter/Filters';

interface NotificationsSearchBarProps {
  startTime: ShortDate;
  setStartTime: (startTime: ShortDate) => void;
  endTime: ShortDate;
  setEndTime: (endTime: ShortDate) => void;
  search: string;
  setSearch: (search: string) => void;
  filters: FilterType[];
  setFilters: (filters: FilterType[]) => void;
  refresh: () => void;
}

export function NotificationsSearchBar(props: NotificationsSearchBarProps) {
  const [query, setQuery] = useState(props.search);

  return (
    <>
      <EuiFlexGroup gutterSize="s" alignItems="center">
        <EuiFlexItem>
          <EuiFieldSearch
            data-test-subj="notifications-search-bar-input"
            fullWidth
            isClearable={false}
            placeholder="Search"
            value={query}
            onChange={(e) => setQuery(e.target.value)}
            onSearch={props.setSearch}
          />
        </EuiFlexItem>
        <EuiFlexItem grow={false}>
          <EuiSuperDatePicker
            start={props.startTime}
            end={props.endTime}
            showUpdateButton={false}
            onTimeChange={(e) => {
              props.setStartTime(e.start);
              props.setEndTime(e.end);
            }}
          />
        </EuiFlexItem>
        <EuiFlexItem grow={false}>
          <EuiButton
            iconType="refresh"
            onClick={() => {
              props.setSearch(query);
              props.refresh();
            }}
          >
            Refresh
          </EuiButton>
        </EuiFlexItem>
      </EuiFlexGroup>
      <EuiSpacer size="s" />
      <Filters filters={props.filters} setFilters={props.setFilters} />
    </>
  );
}
