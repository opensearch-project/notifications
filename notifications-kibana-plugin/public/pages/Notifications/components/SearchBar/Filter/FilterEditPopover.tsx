/*
 * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import {
  EuiButton,
  EuiButtonEmpty,
  EuiComboBox,
  EuiComboBoxOptionOption,
  EuiFlexGroup,
  EuiFlexItem,
  EuiFormRow,
  EuiSpacer,
} from '@elastic/eui';
import React, { useMemo, useState } from 'react';
import {
  FilterOperatorType,
  FilterFieldType,
  getFilterOperatorOptions,
  getOperatorString,
  getFilterFieldOptions,
  getValueComponent,
} from '../utils/filterHelpers';
import { FilterType } from './Filters';

interface FilterEditPopoverProps {
  filter?: FilterType;
  closePopover: () => void;
  index?: number;
  setFilter: (newFilter: FilterType, index?: number) => void;
}

export function FilterEditPopover(props: FilterEditPopoverProps) {
  const [selectedFieldOptions, setSelectedFieldOptions] = useState<
    Array<EuiComboBoxOptionOption<string>>
  >(props.filter ? [{ label: props.filter.field }] : []);
  const [selectedOperatorOptions, setSelectedOperatorOptions] = useState<
    Array<EuiComboBoxOptionOption<string>>
  >(
    props.filter
      ? [
          {
            label: getOperatorString(
              props.filter.operator,
              props.filter.inverted
            ),
          },
        ]
      : []
  );
  const [filterValue, setFilterValue] = useState<FilterType['value']>(
    props.filter?.value || []
  );

  const filterFieldOptions = useMemo(getFilterFieldOptions, []);

  return (
    <div style={{ width: 400 }}>
      {/* invisible button workaround to prevent auto focus on context menu panel switch */}
      <button
        style={{
          width: 0,
          height: 0,
          position: 'fixed',
          marginLeft: -10000,
          bottom: 0,
        }}
      />
      <EuiFlexGroup gutterSize="s">
        <EuiFlexItem grow={6}>
          <EuiFormRow label="Field">
            <EuiComboBox
              placeholder="Select a field first"
              isClearable={false}
              options={filterFieldOptions}
              selectedOptions={selectedFieldOptions}
              onChange={(e) => {
                setSelectedFieldOptions(e);
                setSelectedOperatorOptions([]);
                setFilterValue(null);
              }}
              singleSelection={{ asPlainText: true }}
            />
          </EuiFormRow>
        </EuiFlexItem>
        <EuiFlexItem grow={5}>
          <EuiFormRow label="Operator">
            <EuiComboBox
              placeholder={
                selectedFieldOptions.length === 0 ? 'Waiting' : 'Select'
              }
              isClearable={false}
              isDisabled={selectedFieldOptions.length === 0}
              options={
                selectedFieldOptions.length === 0
                  ? []
                  : getFilterOperatorOptions(
                      selectedFieldOptions[0].label as FilterFieldType
                    )
              }
              selectedOptions={selectedOperatorOptions}
              onChange={(e) => {
                setSelectedOperatorOptions(e);
                setFilterValue(null);
              }}
              singleSelection={{ asPlainText: true }}
            />
          </EuiFormRow>
        </EuiFlexItem>
      </EuiFlexGroup>
      {selectedOperatorOptions.length > 0 &&
        getValueComponent(
          selectedFieldOptions[0].label as FilterFieldType,
          selectedOperatorOptions[0].label as FilterOperatorType,
          filterValue,
          setFilterValue
        )}
      <EuiSpacer size="m" />
      <EuiFlexGroup gutterSize="s" justifyContent="flexEnd">
        <EuiFlexItem grow={false}>
          <EuiButtonEmpty
            data-test-subj="filter-popover-cancel-button"
            onClick={props.closePopover}
          >
            Cancel
          </EuiButtonEmpty>
        </EuiFlexItem>
        <EuiFlexItem grow={false}>
          <EuiButton
            fill
            disabled={
              selectedFieldOptions.length === 0 ||
              selectedOperatorOptions.length === 0 ||
              !(filterValue && filterValue.length > 0)
            }
            onClick={() => {
              props.closePopover();
              props.setFilter(
                {
                  field: selectedFieldOptions[0].label as FilterFieldType,
                  operator: selectedOperatorOptions[0]
                    .label as FilterOperatorType,
                  value: filterValue,
                  inverted: selectedOperatorOptions[0].label.includes('not'),
                  disabled: props.filter ? props.filter.disabled : false,
                },
                props.index
              );
            }}
          >
            Save
          </EuiButton>
        </EuiFlexItem>
      </EuiFlexGroup>
    </div>
  );
}
