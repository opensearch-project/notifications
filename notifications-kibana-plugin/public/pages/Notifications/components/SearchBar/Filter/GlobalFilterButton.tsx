/*
 * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import {
  EuiButtonIcon,
  EuiContextMenu,
  EuiIcon,
  EuiPopover,
} from '@elastic/eui';
import React, { useState } from 'react';
import { FilterType } from './Filters';

interface GlobalFilterButtonProps {
  filters: FilterType[];
  setFilters: (filters: FilterType[]) => void;
}

export function GlobalFilterButton(props: GlobalFilterButtonProps) {
  const [isPopoverOpen, setIsPopoverOpen] = useState(false);
  const globalPopoverPanels = [
    {
      id: 0,
      title: 'Change all filters',
      items: [
        {
          name: 'Enable all',
          icon: <EuiIcon type="eye" size="m" />,
          onClick: () => {
            props.setFilters(
              props.filters.map((filter) => ({ ...filter, disabled: false }))
            );
            setIsPopoverOpen(false);
          },
        },
        {
          name: 'Disable all',
          icon: <EuiIcon type="eyeClosed" size="m" />,
          onClick: () => {
            props.setFilters(
              props.filters.map((filter) => ({ ...filter, disabled: true }))
            );
            setIsPopoverOpen(false);
          },
        },
        {
          name: 'Invert inclusion',
          icon: <EuiIcon type="invert" size="m" />,
          onClick: () => {
            props.setFilters(
              props.filters.map((filter) => ({
                ...filter,
                inverted: !filter.inverted,
              }))
            );
            setIsPopoverOpen(false);
          },
        },
        {
          name: 'Invert enabled/disabled',
          icon: <EuiIcon type="eye" size="m" />,
          onClick: () => {
            props.setFilters(
              props.filters.map((filter) => ({
                ...filter,
                disabled: !filter.disabled,
              }))
            );
            setIsPopoverOpen(false);
          },
        },
        {
          name: 'Remove all',
          icon: <EuiIcon type="trash" size="m" />,
          onClick: () => {
            props.setFilters([]);
            setIsPopoverOpen(false);
          },
        },
      ],
    },
  ];

  return (
    <EuiPopover
      isOpen={isPopoverOpen}
      closePopover={() => setIsPopoverOpen(false)}
      button={
        <EuiButtonIcon
          onClick={() => setIsPopoverOpen(true)}
          iconType="filter"
          title="Change all filters"
          aria-label="Change all filters"
        />
      }
      anchorPosition="rightUp"
      panelPaddingSize="none"
      withTitle
    >
      <EuiContextMenu initialPanelId={0} panels={globalPopoverPanels} />
    </EuiPopover>
  );
}
