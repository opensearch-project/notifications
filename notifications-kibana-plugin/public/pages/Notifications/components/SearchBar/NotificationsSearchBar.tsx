/*
 * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import {
  EuiButton,
  EuiFieldSearch,
  EuiFlexGroup,
  EuiFlexItem,
  EuiSpacer,
  EuiSuperDatePicker,
  ShortDate,
} from '@elastic/eui';
import React from 'react';
import { Filters, FilterType } from './Filter/Filters';

interface NotificationsSearchBarProps {
  startTime: ShortDate;
  setStartTime: (startTime: ShortDate) => void;
  endTime: ShortDate;
  setEndTime: (endTime: ShortDate) => void;
  search: string;
  setSearch: (search: string) => void;
  filters: FilterType[];
  setFilters: (filters: FilterType[]) => void;
  refresh: () => void;
}

export function NotificationsSearchBar(props: NotificationsSearchBarProps) {
  return (
    <>
      <EuiFlexGroup gutterSize="s" alignItems="center">
        <EuiFlexItem>
          <EuiFieldSearch
            fullWidth
            isClearable={false}
            placeholder="Search"
            value={props.search}
            onChange={(e) => props.setSearch(e.target.value)}
            onSearch={props.refresh}
          />
        </EuiFlexItem>
        <EuiFlexItem grow={false}>
          <EuiSuperDatePicker
            start={props.startTime}
            end={props.endTime}
            showUpdateButton={false}
            onTimeChange={(e) => {
              props.setStartTime(e.start);
              props.setEndTime(e.end);
            }}
          />
        </EuiFlexItem>
        <EuiFlexItem grow={false}>
          <EuiButton size="s" onClick={props.refresh}>
            Refresh
          </EuiButton>
        </EuiFlexItem>
      </EuiFlexGroup>
      <EuiSpacer size="s" />
      <Filters filters={props.filters} setFilters={props.setFilters} />
    </>
  );
}
