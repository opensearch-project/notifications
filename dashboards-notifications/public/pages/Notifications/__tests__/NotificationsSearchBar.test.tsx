/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { fireEvent, render } from '@testing-library/react';
import React from 'react';
import { FilterType } from '../component/SearchBar/Filter/Filters';
import { NotificationsSearchBar } from '../component/SearchBar/NotificationsSearchBar';

describe('<NotificationsSearchBar /> spec', () => {
  it('renders the component', () => {
    const filters: FilterType[] = [
      {
        field: 'Channel',
        operator: 'is',
        value: 'test',
        inverted: false,
        disabled: true,
      },
      {
        field: 'Source',
        operator: 'is not one of',
        value: [{ label: 'Alerting' }, { label: 'ISM' }],
        inverted: true,
        disabled: false,
      },
    ];
    const setStartTime = jest.fn();
    const setEndTime = jest.fn();
    const setSearch = jest.fn();
    const setFilters = jest.fn();
    const refresh = jest.fn();
    const utils = render(
      <NotificationsSearchBar
        startTime="now-1d"
        setStartTime={setStartTime}
        endTime="now"
        setEndTime={setEndTime}
        search="test"
        setSearch={setSearch}
        filters={filters}
        setFilters={setFilters}
        refresh={refresh}
      />
    );
    expect(utils.container.firstChild).toMatchSnapshot();

    const searchBar = utils.getByPlaceholderText('Search');
    fireEvent.change(searchBar, { target: { value: 'new query' } });
    expect(setSearch).toBeCalledWith('new query');

    utils.getByText('Refresh').click();
    expect(refresh).toBeCalled();
  });
});
