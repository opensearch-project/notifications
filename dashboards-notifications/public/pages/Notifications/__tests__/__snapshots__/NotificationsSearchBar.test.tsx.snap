// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NotificationsSearchBar /> spec renders the component 1`] = `
<div
  class="euiFlexGroup euiFlexGroup--gutterSmall euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--responsive"
>
  <div
    class="euiFlexItem"
  >
    <div
      class="euiFormControlLayout euiFormControlLayout--fullWidth"
    >
      <div
        class="euiFormControlLayout__childrenWrapper"
      >
        <input
          class="euiFieldSearch euiFieldSearch--fullWidth"
          placeholder="Search"
          type="search"
        />
        <div
          class="euiFormControlLayoutIcons"
        >
          <span
            class="euiFormControlLayoutCustomIcon"
          >
            <svg
              aria-hidden="true"
              class="euiIcon euiIcon--medium euiIcon-isLoading euiFormControlLayoutCustomIcon__icon"
              focusable="false"
              height="16"
              role="img"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            />
          </span>
        </div>
      </div>
    </div>
  </div>
  <div
    class="euiFlexItem euiFlexItem--flexGrowZero"
  >
    <div
      class="euiFlexGroup euiFlexGroup--gutterSmall euiFlexGroup--directionRow euiSuperDatePicker__flexWrapper euiSuperDatePicker__flexWrapper--noUpdateButton"
    >
      <div
        class="euiFlexItem"
      >
        <div
          class="euiFormControlLayout euiFormControlLayout--group euiSuperDatePicker"
        >
          <div
            class="euiPopover euiPopover--anchorDownLeft"
            id="QuickSelectPopover"
          >
            <div
              class="euiPopover__anchor euiQuickSelectPopover__anchor"
            >
              <button
                aria-label="Date quick select"
                class="euiButtonEmpty euiButtonEmpty--primary euiButtonEmpty--xSmall euiFormControlLayout__prepend"
                data-test-subj="superDatePickerToggleQuickMenuButton"
                type="button"
              >
                <span
                  class="euiButtonContent euiButtonContent--iconRight euiButtonEmpty__content"
                >
                  <svg
                    aria-hidden="true"
                    class="euiIcon euiIcon--medium euiIcon-isLoading euiButtonContent__icon"
                    focusable="false"
                    height="16"
                    role="img"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  />
                  <span
                    class="euiButtonEmpty__text euiQuickSelectPopover__buttonText"
                  >
                    <svg
                      aria-hidden="true"
                      class="euiIcon euiIcon--medium euiIcon-isLoading"
                      focusable="false"
                      height="16"
                      role="img"
                      viewBox="0 0 16 16"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    />
                  </span>
                </span>
              </button>
            </div>
          </div>
          <div
            class="euiFormControlLayout__childrenWrapper"
          >
            <div
              class="euiDatePickerRange euiDatePickerRange--inGroup"
            >
              <button
                class="euiSuperDatePicker__prettyFormat"
                data-test-subj="superDatePickerShowDatesButton"
              >
                Last 1 day
                <span
                  class="euiSuperDatePicker__prettyFormatLink"
                >
                  Show dates
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="euiFlexItem euiFlexItem--flexGrowZero"
  >
    <button
      class="euiButton euiButton--primary"
      type="button"
    >
      <span
        class="euiButtonContent euiButton__content"
      >
        <svg
          aria-hidden="true"
          class="euiIcon euiIcon--medium euiIcon-isLoading euiButtonContent__icon"
          focusable="false"
          height="16"
          role="img"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        />
        <span
          class="euiButton__text"
        >
          Refresh
        </span>
      </span>
    </button>
  </div>
</div>
`;
